open_spiel/python/games/domino.py

```diff
@@ -0,0 +1,346 @@
+
+            self.hands[0].sort()
+            self.hands[1].sort()
+
+            self._next_player = 0
+            # calc all possible move for the first player to play
+            self.player_legal_actions = self.get_legal_actions(self._next_player)
+        else:
+            action = self.player_legal_actions[action]
+            self.gameHistory.append(action)
+            my_idx = action.player
+            my_hand = self.hands[my_idx]
+            my_hand.remove(action.tile_to_put)
+            self.open_edges = action.new_edges
+
+            if not my_hand:
+                self._game_over = True  # player played his last tile
+                return
+
+            opp_idx = 1 - my_idx
+            opp_legal_actions = self.get_legal_actions(opp_idx)
+
+            if opp_legal_actions:
+                self._next_player = opp_idx
+                self.player_legal_actions = opp_legal_actions
+                return
+
+            my_legal_actions = self.get_legal_actions(my_idx)
+            if my_legal_actions:
+                self._next_player = my_idx
+                self.player_legal_actions = my_legal_actions
+                return
+
+            self._game_over = True  # both players are blocked
+
+    def _action_to_string(self, player, action):
+        """Action -> string."""
+        if player == pyspiel.PlayerId.CHANCE:
+            return f"Deal {self.deck[action]}"
+        return str(self.player_legal_actions[action])
+
+    def is_terminal(self):
+        """Returns True if the game is over."""
+        return self._game_over
+
+    def returns(self):
+        """Total reward for each player over the course of the game so far."""
+
+        if not self.is_terminal():
+            return [0, 0]
+
+        sum_of_pips0 = sum(t[0] + t[1] for t in self.hands[0])
+        sum_of_pips1 = sum(t[0] + t[1] for t in self.hands[1])
+
+        if sum_of_pips1 == sum_of_pips0:
+            return [0, 0]
+
+        if sum_of_pips1 > sum_of_pips0:
+            return [sum_of_pips1, -sum_of_pips1]
+        return [-sum_of_pips0, sum_of_pips0]
+
+    def __str__(self):
+        """String for debug purposes. No particular semantics are required."""
+        hand0 = [str(c) for c in self.hands[0]]
+        hand1 = [str(c) for c in self.hands[1]]
+        history = [str(a) for a in self.gameHistory]
+        s = f'hand0:{hand0} hand1:{hand1} history:{history}'
+        return s
+
+
+class DominoObserver:
+    """Observer, conforming to the PyObserver interface (see observation.py)."""
+
+    def __init__(self, iig_obs_type, params):
+        """Initializes an empty observation tensor."""
+        if params:
+            raise ValueError(f"Observation parameters not supported; passed {params}")
+
+        # Determine which observation pieces we want to include.
+        pieces = [("player", 2, (2,))]
+
+        if iig_obs_type.private_info == pyspiel.PrivateInfoType.SINGLE_PLAYER:
```

Comment:
Can you add some comments to explain the numbers here?

Reply:
Please fix